cpp/src/community/louvain.cuh

```diff
@@ -0,0 +1,639 @@
+                                           weight_t resolution)
+  {
+    int num_level{0};
+
+    weight_t total_edge_weight =
+      thrust::reduce(rmm::exec_policy(stream_)->on(stream_), weights_v_.begin(), weights_v_.end());
+
+    weight_t best_modularity = weight_t{-1};
+
+    //
+    //  Initialize every cluster to reference each vertex to itself
+    //
+    thrust::sequence(rmm::exec_policy(stream_)->on(stream_), cluster_v_.begin(), cluster_v_.end());
+    thrust::copy(
+      rmm::exec_policy(stream_)->on(stream_), cluster_v_.begin(), cluster_v_.end(), d_cluster_vec);
+
+    //
+    //  Our copy of the graph.  Each iteration of the outer loop will
+    //  shrink this copy of the graph.
+    //
+    GraphCSRView<vertex_t, edge_t, weight_t> current_graph(offsets_v_.data().get(),
+                                                           indices_v_.data().get(),
+                                                           weights_v_.data().get(),
+                                                           number_of_vertices_,
+                                                           number_of_edges_);
+
+    current_graph.get_source_indices(src_indices_v_.data().get());
+
+    while (num_level < max_level) {
+      compute_vertex_and_cluster_weights(current_graph);
+
+      weight_t new_Q = update_clustering(total_edge_weight, resolution, current_graph);
+
+      if (new_Q <= best_modularity) { break; }
+
+      best_modularity = new_Q;
+
+      shrink_graph(current_graph, d_cluster_vec);
+
+      num_level++;
+    }
+
+    timer_display(std::cout);
+
+    return std::make_pair(num_level, best_modularity);
+  }
+
+ protected:
+  void timer_start(std::string const &region)
+  {
+#ifdef TIMING
+    hr_timer_.start(region);
+#endif
+  }
+
+  void timer_stop(cudaStream_t stream)
+  {
+#ifdef TIMING
+    CUDA_TRY(cudaStreamSynchronize(stream));
+    hr_timer_.stop();
+#endif
+  }
+
+  void timer_display(std::ostream &os)
+  {
+#ifdef TIMING
+    hr_timer_.display(os);
+#endif
+  }
+
+ public:
+  void compute_vertex_and_cluster_weights(graph_type const &graph)
+  {
+    timer_start("compute_vertex_and_cluster_weights");
+
+    edge_t const *d_offsets     = graph.offsets;
+    vertex_t const *d_indices   = graph.indices;
+    weight_t const *d_weights   = graph.edge_data;
+    weight_t *d_vertex_weights  = vertex_weights_v_.data().get();
+    weight_t *d_cluster_weights = cluster_weights_v_.data().get();
+
+    //
+    // MNMG:  copy_v_transform_reduce_out_nbr, then copy
+    //
+    thrust::for_each(
+      rmm::exec_policy(stream_)->on(stream_),
+      thrust::make_counting_iterator<edge_t>(0),
+      thrust::make_counting_iterator<edge_t>(graph.number_of_vertices),
+      [d_offsets, d_indices, d_weights, d_vertex_weights, d_cluster_weights] __device__(
+        vertex_t src) {
+        weight_t sum{0.0};
+
+        for (edge_t i = d_offsets[src]; i < d_offsets[src + 1]; ++i) { sum += d_weights[i]; }
```

Comment:
Please consider using `thrust::reduce(thrust::seq,...)` with `__host__ __device__ make_counting_iterator()` instead of a for loop (and still avoid dynamic parallelism).

Reply:
Done in next commit